### ISSUE TITLE
variable name not replaced in unreachable code causes compiler error with applied fixes

### ISSUE THREAD

**Issue by antonilol** [rust-lang/rust-clippy#11384]
### Summary

a variable used only after a `todo!();` was marked unused. when applying automatic fixes only the parameter name got changed to start with an underscore, causing an error because the original name was still after the `todo!();`

full clippy output (`cargo clippy --fix --allow-no-vcs`):

```
    Checking clippy_bug v0.1.0 (/tmp/clippy_bug)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_bug`

after fixes were automatically applied the compiler reported errors within these files:

  * /home/antoni/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/macros/mod.rs
  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0425]: cannot find value `value` in this scope
 --> src/main.rs:7:5
  |
7 |     value
  |     ^^^^^ help: a local variable with a similar name exists: `_value`

warning: unreachable expression
 --> src/main.rs:7:5
  |
6 |     todo!();
  |     ------- any code following this expression is unreachable
7 |     value
  |     ^^^^^ unreachable expression
  |
  = note: `#[warn(unreachable_code)]` on by default

error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0425`.
Original diagnostics will follow.

warning: unreachable expression
 --> src/main.rs:7:5
  |
6 |     todo!();
  |     ------- any code following this expression is unreachable
7 |     value
  |     ^^^^^ unreachable expression
  |
  = note: `#[warn(unreachable_code)]` on by default

warning: unused variable: `value`
 --> src/main.rs:5:13
  |
5 | fn function(value: usize) -> usize {
  |             ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: function `function` is never used
 --> src/main.rs:5:4
  |
5 | fn function(value: usize) -> usize {
  |    ^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: `clippy_bug` (bin "clippy_bug") generated 3 warnings (run `cargo clippy --fix --bin "clippy_bug"` to apply 1 suggestion)
warning: `clippy_bug` (bin "clippy_bug" test) generated 3 warnings (3 duplicates)
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
```

### Reproducer

I tried this code:

```rust
fn main() {
    println!("Hello, world!");
}

fn function(value: usize) -> usize {
    todo!();
    value
}
```

I expected to see this happen:

both instances of `value` got replaced with `_value`, or none at all

Instead, this happened:

one of them got replaced, resulting in ``error[E0425]: cannot find value `value` in this scope``

### Version

```text
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5
```

**Comment by y21** [comment]
I don't think this is an issue with clippy. `unreachable_code` + `unused_variables` are rustc lints, so this should probably be on the rust-lang/rust repo.
This can also be reproduced with just `cargo fix`

**Comment by y21** [comment]
To add to my comment, this seems to be a known bug: https://github.com/rust-lang/rust/issues/107734

**Comment by antonilol** [comment]
right, i see, closing as duplicate then
